src/option_learning.py

```diff
@@ -212,3 +220,206 @@ def update_segment_from_option_spec(self, segment: Segment,
+        action_arr = self._regressor.predict(x)
+        if np.isnan(action_arr).any():
+            raise ApproachFailure("Option policy returned nan.")
+        return Action(np.array(action_arr, dtype=np.float32))
+
+    def _effect_based_terminal(self, state: State, memory: Dict,
+                               objects: Sequence[Object],
+                               params: Array) -> bool:
+        del memory, params  # unused
+        # The hope is that we terminate in the effects.
+        grounded_op = self._operator.ground(tuple(objects))
+        abs_state = utils.abstract(state, self._predicates)
+        if grounded_op.add_effects.issubset(abs_state) and \
+            not grounded_op.delete_effects & abs_state:
+            return True
+        # If we fall outside of the space where the preconditions hold, also
+        # terminate. The assumption is that the option should take us from one
+        # abstract state to another without hitting any others in between.
+        if not grounded_op.preconditions.issubset(abs_state):
+            return True
+        # Not yet done.
+        return False
+
+
+class _SimpleOptionLearner(_OptionLearnerBase):
+    """The option learner that learns options from fixed-length input vectors.
+
+    That is, the the input data only includes objects whose state
+    changes from the first state in the segment to the last state in the
+    segment. The input data does not include data from other objects in
+    the scene, where the number of other objects would differ across
+    different instiantiations of the env.
+    """
+
+    def __init__(self) -> None:
+        super().__init__()
+        # While learning the policy, we record the map from each segment to
+        # the option parameterization, so we don't need to recompute it in
+        # update_segment_from_option_spec.
+        self._segment_to_grounding: Dict[Segment, Tuple[Sequence[Object],
+                                                        Array]] = {}
+
+    def learn_option_specs(self, strips_ops: List[STRIPSOperator],
+                           datastores: List[Datastore]) -> List[OptionSpec]:
+        # In this paradigm, the option initiable and termination are determined
+        # from the operators, so the main thing that needs to be learned is the
+        # option policy.
+        option_specs: List[Tuple[ParameterizedOption, List[Variable]]] = []
+
+        assert len(strips_ops) == len(datastores)
+
+        for op, datastore in zip(strips_ops, datastores):
+            print(f"\nLearning option for NSRT {op.name}")
+
+            X_regressor: List[Array] = []
+            Y_regressor = []
+
+            # The superset of all objects whose states we may include in the
+            # option params is op.parameters. But we only want to include
+            # those objects that have state changes (in at least some data).
+            changing_parameter_set = self._get_changing_parameters(datastore)
```

Comment:
Why? Is this just an efficiency thing or an absolute necessity?

Reply:
it's a learning efficiency thing, I'll indicate this in the comment